hack/test-go.sh

```diff
@@ -1,129 +1,259 @@
-  fi
+# Build arguments for 'go test'
+if [[ -z "${verbose}" && -n "${junit_report}" ]]; then
+    # verbosity can be set explicitly by the user or set implicitly by asking for the jUnit 
+    # XML report, so we only want to add the flag if it hasn't been added by a user already 
+    # and is being implicitly set by jUnit report generation
+    gotest_flags="${gotest_flags} -v"
 fi
 
-if [ -z ${KUBE_RACE+x} ]; then
-  KUBE_RACE=""
+if [[ -n "${detect_races}" ]]; then
+    gotest_flags="${gotest_flags} -race"
 fi
 
-KUBE_TIMEOUT=${KUBE_TIMEOUT:-'60s'}
-
-if [ "${1-}" != "" ]; then
-  suffix=/...
-  if [[ "${1}" == *"${suffix}" ]]; then
-    # if the package name ends with /... they are intending a recursive test
-    test_packages=$(find_test_dirs "${1%$suffix}")
-  else
-    test_packages="$OS_GO_PACKAGE/$1"
-  fi
-elif [ -n "$TEST_KUBE" ]; then
-  test_packages=`find_test_dirs "."; special_upstream_test_dirs; find_upstream_test_dirs`
+# check to see if user has not disabled coverage mode
+if [[ -n "${coverage_spec}" ]]; then
+    # if we have a coverage spec set, we add it. '-race' implies '-cover -covermode atomic'
+    # but specifying both at the same time does not lead to an error so we can add both specs
+    gotest_flags="${gotest_flags} ${coverage_spec}"
+fi
+
+# check to see if user has not disabled test timeouts
+if [[ -n "${test_timeout}" ]]; then
+    gotest_flags="${gotest_flags} -timeout ${test_timeout}"
+fi
+
+# list_test_packages_under lists all packages containing Golang test files that we want to run as unit tests
+# under the given base dir in the OpenShift Origin tree
+function list_test_packages_under() {
+    local basedir=$@
+
+    # we do not quote ${basedir} to allow for multiple arguments to be passed in as well as to allow for
+    # arguments that use expansion, e.g. paths containing brace expansion or wildcards
+    find ${basedir} -not \(                   \
+        \(                                    \
+              -path 'Godeps'                  \
+              -o -path '*_output'             \
+              -o -path '*_tools'              \
+              -o -path '*.git'                \
+              -o -path '*openshift.local.*'   \
+              -o -path '*Godeps/*'            \
+              -o -path '*assets/node_modules' \
+              -o -path '*test/*'              \
+        \) -prune                             \
+    \) -name '*_test.go' -print0 | xargs -0n1 dirname -z | sort -uz | xargs -0n1 printf "${OS_GO_PACKAGE}/%s\n"
+}
+
+# Break up the positional arguments into packages that need to be tested and arguments that need to be passed to `go test`
+first_flag_arg=-1
+for (( i = 1; i < "$#"; i++ )); do
+    if [[ "${!i:0:1}" = "-" ]]; then
+        # we found an arg that begins with a dash, so we stop interpreting arguments 
+        # henceforth as packages and instead interpret them as flags to give to `go test`
+        first_flag_arg=$i
+        break
+    fi
+done
+
+package_args=
```

Comment:
This line is unnecessary (but maybe our tentative style guidelines require it?).